lib/ipam/ipam.go

```diff
@@ -1383,6 +1349,69 @@ func (c ipamClient) convertBackendToIPAMConfig(cfg *model.IPAMConfig) *IPAMConfi
 	}
 }
 
+// ensureConsistentAffinity retrieves the pool and node for the given block and determines
+// if the pool still selects node. If it no longer matches, it will release the block
+// affinity for that node.
+// Returns a bool indicating if the block affinity was released.
+func (c ipamClient) ensureConsistentAffinity(ctx context.Context, b *model.AllocationBlock) (bool, error) {
+
+	// Retrieve node for this allocation. We do this so we can clean up affinity for blocks
+	// which should no longer be affine to this host.
+	host := getHostAffinity(b)
+	logCtx := log.WithFields(log.Fields{"cidr": b.CIDR, "host": host})
+
+	// If no hostname is found on the block affinity,
+	// there is no need to do an ip pool node selection check.
+	if host == "" {
+		logCtx.Debug("Block already has no affinity")
+		return false, nil
+	}
+
+	// If the IP pool which owns this block no longer selects this node,
+	// we should release the block's affinity to this node so it can be
+	// used elsewhere.
+	logCtx.Debugf("Looking up node labels for host affinity")
+	node, err := c.client.Get(ctx, model.ResourceKey{Kind: v3.KindNode, Name: host}, "")
+	if err != nil {
+		if _, ok := err.(cerrors.ErrorResourceDoesNotExist); !ok {
+			logCtx.WithError(err).WithField("node", host).Error("Failed to get node for host")
+			return false, err
+		}
+		logCtx.Info("Node doesn't exist, releasing IPs anyway")
+		return false, nil
+	}
+
+	// Make sure the returned value is a valid node.
+	v3n, ok := node.Value.(*v3.Node)
+	if !ok {
+		return false, fmt.Errorf("Datastore returned malformed node object")
+	}
+
+	// Fetch the pool for the given CIDR and check if it selects the node.
+	if pool := c.blockReaderWriter.getPoolForIP(cnet.IP{b.CIDR.IPNet.IP}); pool == nil {
+		logCtx.Debug("no pools own this block")
+		return false, nil
+	} else {
+		logCtx = logCtx.WithField("pool", pool)
```

Comment:
I don't think we want to log the ensure IP pool in the context logger. We can do it once as a debug log, or maybe just log out the CIDR and selector since that's really what we care about?

Reply:
Probably just use the existing `logCtx` and write out something like this:

```
logCtx.Debugf("Block belongs with IPPool %s with nodeSelector %s", ...)
```

Reply:
Yeah, logging out the entire IP pool is overkill. The CIDR context is already added at the beginning of the function so I'll just add the selector.